File: src/com/leetcode/Hard/MergeKSortedLists.java
package com.leetcode.Hard;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

// 23. Merge k Sorted Lists
public class MergeKSortedLists {
    public static class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0)
            return null;

        List<Integer> list = new ArrayList<>();
        for (ListNode node : lists) {
            ListNode curr = node;
            while (curr != null) {
                list.add(curr.val);
                curr = curr.next;
            }
        }

        Collections.sort(list);
        ListNode temp = new ListNode(0);
        ListNode curr = temp;
        for (int i : list) {
            ListNode node = new ListNode(i);
            curr = curr.next;
        }
        return temp.next;
    }
}

File: JavaScript/Hard/MergeKSortedLists.js
/**
 * 23. Merge k Sorted Lists
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
const mergeKLists = function (lists) {
  if (lists.length === 0) return null;

  while (lists.length > 1) {
    let l1 = lists.shift();
    let l2 = lists.shift();
    const merged = mergeTwoLists(l1, l2);
    lists.push(merged);
  }

  return lists[0];
};

const mergeTwoLists = function (l1, l2) {
  if (!l1) return l2;
  if (!l2) return l1;

  if (l1.val < l2.val) {
    l1.next = mergeTwoLists(l1.next, l2);
    return l1;
  } else {
    l2.next = mergeTwoLists(l1, l2.next);
    return l2;
  }
};
